numba/targets/hashing.py

```diff
@@ -0,0 +1,605 @@
+        tupty, = signature.args
+        tup, = args
+        lty = cgctx.get_value_type(signature.return_type)
+        x = ir.Constant(lty, 0x345678)
+        mult = ir.Constant(lty, _PyHASH_MULTIPLIER)
+        shift = ir.Constant(lty, 82520)
+        tl = len(tupty)
+        for i, packed in enumerate(zip(tupty.types, range(tl - 1, -1, -1))):
+            ty, l = packed
+            sig = fnty.get_call_type(tyctx, (ty,), {})
+            impl = cgctx.get_function(fnty, sig)
+            tuple_val = builder.extract_value(tup, i)
+            y = impl(builder, (tuple_val,))
+            xxory = builder.xor(x, y)
+            x = builder.mul(xxory, mult)
+            lconst = ir.Constant(lty, l)
+            mult = builder.add(mult, shift)
+            mult = builder.add(mult, lconst)
+            mult = builder.add(mult, lconst)
+        x = builder.add(x, ir.Constant(lty, 97531))
+        return x
+    sig = _Py_hash_t(val)
+    return sig, impl
+
+
+@overload_method(types.BaseTuple, '__hash__')
+def tuple_hash(val):
+    if isinstance(val, types.Sequence):
+        def impl(val):
+            return _tuple_hash(val)
+        return impl
+    else:
+        def impl(val):
+            hashed = _Py_hash_t(_tuple_hash_resolve(val))
+            return process_return(hashed)
+        return impl
+
+
+# ------------------------------------------------------------------------------
+# String/bytes hashing needs hashseed info, this is from:
+# https://stackoverflow.com/a/41088757
+# with thanks to Martijn Pieters
+from ctypes import (  # noqa
+    c_size_t,
+    c_ubyte,
+    c_uint64,
+    pythonapi,
+    Structure,
+    Union,
+)  # noqa
+
+
+class FNV(Structure):
+    _fields_ = [
+        ('prefix', c_size_t),
+        ('suffix', c_size_t)
+    ]
+
+
+class SIPHASH(Structure):
+    _fields_ = [
+        ('k0', c_uint64),
+        ('k1', c_uint64),
+    ]
+
+
+class DJBX33A(Structure):
+    _fields_ = [
+        ('padding', c_ubyte * 16),
+        ('suffix', c_size_t),
+    ]
+
+
+class EXPAT(Structure):
+    _fields_ = [
+        ('padding', c_ubyte * 16),
+        ('hashsalt', c_size_t),
+    ]
+
+
+class _Py_HashSecret_t(Union):
+    _fields_ = [
+        # ensure 24 bytes
+        ('uc', c_ubyte * 24),
+        # two Py_hash_t for FNV
+        ('fnv', FNV),
+        # two uint64 for SipHash24
+        ('siphash', SIPHASH),
+        # a different (!) Py_hash_t for small string optimization
+        ('djbx33a', DJBX33A),
+        ('expat', EXPAT),
+    ]
+
+
+# Only a few members are needed at present
+_Py_hashSecret = _Py_HashSecret_t.in_dll(pythonapi, '_Py_HashSecret')
+_Py_HashSecret_djbx33a = _Py_hashSecret.djbx33a
+_Py_HashSecret_djbx33a_suffix = _Py_hashSecret.djbx33a.suffix
+_Py_HashSecret_siphash_k0 = _Py_hashSecret.siphash.k0
+_Py_HashSecret_siphash_k1 = _Py_hashSecret.siphash.k1
+# ------------------------------------------------------------------------------
+
+
+@intrinsic
+def grabbyte(typingctx, data, offset):
```

Comment:
This is useful and generic.  Can you add this under `numba.unsafe`?

Reply:
done.